feat: make ConnectivityListener configurable, and remove unnecessary network availability checks

gkc opened the pull request:
Closes #1099

See analysis [here](https://github.com/atsign-foundation/sshnoports/issues/292#issuecomment-1656710825) which led to the changes in this PR

- Eliminate direct use of InternetConnectivityChecker from at_client package except for a method in RemoteSecondary which (a) is unused and now marked `@Deprecated` and where (b) the usage is non-problematic (checks if atServer's host and port are reachable)
- Eliminate wrapper class at_client.NetworkUtil and all its usages in at_client
- Eliminate indirect (via at_client.ConnectivityListener) use of InternetConnectivityChecker from at_client package
- For backwards compatibility, still exports ConnectivityListener  (which is in use by multiple programs including sshnpd), but
  - (a) adds optional constructor parameters which programs can use if they wish to continue to use ConnectivityListener
  - (b) ConnectivityListener is marked deprecated as connectivity listening is not functionality which at_client should export

**- What I did**
- [x] Deprecate `RemoteSecondary.isAvailable()`
- [x] Remove Monitor's unnecessary use of NetworkUtil.isNetworkAvailable
- [x] Remove unnecessary use of ConnectivityListener from NotificationServiceImpl
- [x] Remove SyncServiceImpl's unnecessary use of NetworkUtil.isNetworkAvailable
- [x] Remove AtClientValidation.validateAtKey's unnecessary use of NetworkUtil.isNetworkAvailable
- [x] Adjust any unit tests which currently use mock network availability checks to simulate atServer reachability to instead use other mock approaches instead, without changing those tests's other setup or `expect`s
- [x] Remove the now-unused NetworkUtil class
- [x] Add hostname, port and a checkInterval as optional parameter to ConnectivityListener constructor and use accordingly
- [x] Deprecate ConnectivityListener class as it's not a core concern of the at_client package
- [x] Some software supply chain management
  - adjusted dependency configuration for the crypto packages we depend on
  - defensive measures to prevent the next release of the encrypt package from causing us problems, as the behaviour of IV.fromLength has changed in the main branch (unpublished)

**- How I did it**

**- How to verify it**
Tests pass

**- Description for the changelog**
feat: Make `ConnectivityListener` configurable by adding optional parameters to its constructor.
feat: Remove unnecessary use of network availability checks in various places

gkc committed:
feat: deprecated RemoteSecondary.isAvailable() method (which is completely unused)

gkc committed:
feat: removed network availability check from Monitor
refactor: refactored Monitor unit tests to depend on ability to create a connection rather than network availability

gkc committed:
feat: removed NotificationServiceImpl's use of ConnectivityListener

gkc committed:
feat: Remove SyncServiceImpl's unnecessary usage of NetworkUtil.isNetworkAvailable

gkc committed:
refactor: Make sync unit tests simulate atServer unreachable by throwing AtConnectExceptions where previously they relied on having calls to a mockNetworkUtil.isNetworkAvailable return false

gkc committed:
feat: remove `AtClientValidation.validateAtKey`'s unnecessary check for existence of an atSign

gkc committed:
feat: remove `AtClientValidation.validateAtKey`'s unnecessary check for existence of an atSign

gkc committed:
refactor: remove unused networkUtil instance variable from SyncServiceImpl (not part of public interface of that class, only used for testing

gkc committed:
refactor: remove unused NetworkUtil class and file network_util.dart

gkc committed:
feat: Add hostname, port and a checkInterval as optional parameters to ConnectivityListener constructor and use accordingly

gkc committed:
style: ran dart format

gkc committed:
build: update CHANGELOG, pubspec and AtClientConfig for 3.0.64

cconstab reviewed (commented):
I hate that connectivity listener uses tcp rather than ICMP

cconstab commented:
Another process wrap but well done  

https://pub.dev/packages/dart_ping


And some discussion on the topic

https://github.com/dart-lang/sdk/issues/33764

gkc commented:
> I hate that connectivity listener uses tcp rather than ICMP

ConnectivityListener is ours; it wraps InternetConnectivityChecker

Neither are being used by at_client any more so 🤷‍♂️

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-fewer-more-configurable-network-checks

# Conflicts:
#	packages/at_client/CHANGELOG.md

gkc commented:
> Another process wrap but well done
> 
> https://pub.dev/packages/dart_ping
> 
> And some discussion on the topic
> 
> [dart-lang/sdk#33764](https://github.com/dart-lang/sdk/issues/33764)

Now that at_client makes no use of it, I'd be happy to mark at_client's ConnectivityListener class as Deprecated and leave it to programs to use whatever they want to instead, since really it's nothing to do with AtClient

gkc committed:
refactor: remove unused stuff from remote_secondary_test.dart
docs: update `@Deprecated` message for `RemoteSecondary.isAvailable`

gkc committed:
build: software supply chain management
- add direct dependency on asn1lib
- pin versions for asnlib, crypton and encrypt packages

gkc committed:
feat: Deprecate ConnectivityListener class

gkc commented:
Checks currently failing because one of our cicd hosts is down

gkc commented:
Checks complete thank you @cpswan

gkc commented on packages/at_client/lib/src/converters/encryption/aes_converter.dart:
```diff
@@ -111,7 +111,8 @@ class AESDecryptionSink extends ChunkedConversionSink<List<int>> {
   if (ivBase64 == null) {
-    return IV.fromLength(16);
+    // From the bad old days when we weren't setting IVs
+    return IV(Uint8List(16));
```
This change is a defensive measure to prevent the next release of the `encrypt` package from causing us problems, as the behaviour of `IV.fromLength` has changed

gkc commented on packages/at_client/lib/src/util/encryption_util.dart:
```diff
@@ -12,7 +12,8 @@ class EncryptionUtil {
     if (ivBase64 == null) {
-      return IV.fromLength(16);
+      // From the bad old days when we weren't setting IVs
+      return IV(Uint8List(16));
```
This change is a defensive measure to prevent the next release of the `encrypt` package from causing us problems, as the behaviour of `IV.fromLength` has changed

cconstab commented:
Everything works as expected thanks @gkc 


<img width="1290" alt="Screenshot 2023-07-30 at 10 05 59" src="https://github.com/atsign-foundation/at_client_sdk/assets/6131216/141ff949-c0b9-4b7c-92e5-095df151b20c">

cconstab reviewed (approved):
Tested by hand as well as via Ci/CD

gkc committed:
Merge branch 'trunk' into gkc-fewer-more-configurable-network-checks
